### ISSUE TITLE
How to use line attributes properly?

### ISSUE THREAD

**Issue by g-borgulya** [ether/etherpad-lite#2709]
I'm not sure if it's OK to post questions like this here.

I'm working on a plugin: I try to add an extra *span* with some class settings to every line that has any of my custom tags in them. In my solution I experience some strange behavior with the caret. I guess I simply don't understand how line attributes work.

I do the following:

1. in **aceEditEvent**, if *context.callstack.docTextChanged* is *true*, I iterate through *rep.lines.atIndex(currentLine).lineNode.children* and if there are any custom tags, I use *documentAttributeManager.setAttributeOnLine* to mark the line as special.
2. I use **aceAttribsToClasses** to convert the new attribute to a class
3. in **aceDomLineProcessLineAttributes** if I find the newly added class, I add the desired *span* and set the *processedMarker* to *true*.

The result looks perfectly. But if I move the caret in the editor into a line like this, it steps one letter left after every keystroke and also at every *idleWorkTimer* event. This also results in writing backwards in if I type in these lines. It seems to be some internal thing, as none of my methods run when I don't type and the caret is still walking left. If I set *processedMarker* to *false* in the 3rd step, the caret stops moving, however this case I have a * rendered at the beginning of the lines with the new attribute set.

Is this the right approach? Do I miss something important? The documentation talks about **aceRegisterBlockElements** but I'm not adding any new custom HTML tags so I guess I don't need that.

**Comment by JohnMcLear** [comment]
See the source code of https://github.com/johnmclear/ep_headings2

Specifically https://github.com/JohnMcLear/ep_headings2/blob/master/static/js/index.js#L10

**Comment by g-borgulya** [comment]
Thank you, John.
However, I still don't see the solution. I created a very simple plugin that shows my problem: https://github.com/g-borgulya/ep_lineattribute-test
If you write anything in etherpad using this plugin, it sets a lineattribute for the currently edited line, which causes the caret to continuously "walk" to the left while it's in the line. I don't know if this is an etherpad bug or if I'm doing something wrong.

**Comment by JohnMcLear** [comment]
Yea you are doing something wrong, please look at the other projects that do this.

**Comment by g-borgulya** [comment]
OK, found the problem.
I wanted to decorate the lines with span elements, but it simply doesn't seem to work either with or without registering span with aceRegisterBlockElements. It seems that one needs to register a new html tag and use it, otherwise there's a problem with the caret.

As soon as I registered for example 'x-whatever' with aceRegisterBlockElements and used aceDomLineProcessLineAttributes to add it as pre/postHTML, it stopped the unwanted behavior of the  caret. 

I updated the simplified plugin mentioned above, so it's basically a very simple plugin to demonstrate styling a line.
https://github.com/g-borgulya/ep_lineattribute-test
